api/apps/api/src/modules/geo-feature-tags/geo-feature-tags.service.ts

```diff
@@ -0,0 +1,53 @@
+import { InjectRepository } from '@nestjs/typeorm';
+import { GeoFeatureTag } from '@marxan-api/modules/geo-feature-tags/geo-feature-tag.api.entity';
+import { Repository } from 'typeorm';
+import { GeoFeature } from '@marxan-api/modules/geo-features/geo-feature.api.entity';
+import { Either, left, right } from 'fp-ts/lib/Either';
+
+export const featureNotFound = Symbol('feature not found');
+export const featureProjectNotRelated = Symbol(
+  'feature and project not related',
+);
+
+export class GeoFeatureTagsService {
+  constructor(
+    @InjectRepository(GeoFeatureTag)
+    private readonly geoFeatureTagsRepo: Repository<GeoFeatureTag>,
+    @InjectRepository(GeoFeature)
+    private readonly geoFeaturesRepo: Repository<GeoFeature>,
+  ) {}
+
+  async updateTagForFeature(
+    projectId: string,
+    featureId: string,
+    tag: string,
+  ): Promise<
+    Either<typeof featureNotFound | typeof featureProjectNotRelated, any>
+  > {
+    const geoFeature = await this.geoFeaturesRepo.findOne({
+      where: { id: featureId },
+    });
+
+    if (!geoFeature) {
+      return left(featureNotFound);
+    }
+    if (geoFeature.projectId !== projectId) {
+      return left(featureProjectNotRelated);
+    }
```

Comment:
here we're using the user-supplied `projectId` without even checking that this is the id of an actual project - but instead of validating this indirectly by comparing the feature's actual `projectId` with the `projectId` supplied by the user, I would suggest to simplify like this

```suggestion
    const geoFeature = await this.geoFeaturesRepo.findOne({
      where: { id: featureId, projectId },
    });

    if (!geoFeature) {
      return left(featureNotFound);
    }
```

and forget about the `projectId` - what we need to know here is whether the given feature exists within the given project, and I think the suggested check should suffice.